index.d.ts

```diff
@@ -14,9 +14,13 @@ interface EggRedisOptions {
   clients?: Record<string, RedisOptions>;
 }
 
+declare class Singleton<T> {
+  get(key: string): T;
+}
+
 declare module 'egg' {
   interface Application {
-    redis: Redis;
+    redis: Redis | Singleton<Redis>;
```

Comment:
Yep, `Redis | Singleton<Redis>` is right for this plugin. But I don't know whether it should be change to this type because it will make breakchange. 

If the code like this.

```typescript
this.app.redis.set('foo', 'bar')
```

It will cause type error when the type of `app.redis` change to `Redis | Singleton<Redis>` from `Redis`, because there is no `set()` type in both `Redis` or `Singleton<Redis>`. This is the first problem.

The second is, the plugin will became a little hard to use because I have to cast type before using in every time.

```typescript
import { Redis } from 'ioredis';

export default class HomeController extends Controller {
  async index() {
    (<Redis>this.app.redis).set('foo', 'bar');
  },
  async index2() {
    (<Redis>this.app.redis).get('foo');
  },
  async index3() {
    (<Redis>this.app.redis).get('foo');
  }
}
```

And the type `Redis` I can import from `ioredis`, but `Singleton<Redis>` seems I have to declare again in my own project.

I still have no idea to resolve these two problem. do you have any idea?

Reply:
Yes, this change is a break change.
I think the problem is cause by api design. Split single and multi to two different objects might be a good choice. 
Otherwise, use `Redis & Singleton<Redis>` is a way to hack it.

Reply:
Singleton now can inport from egg

Reply:
OK，It seems the only way to compatible.

btw, I think you can submit a PR for adding the declaration of `Singleton` to https://github.com/eggjs/egg , so it can be used by the other plugins.